File: rust/crates/corlinman-channels/src/channel.rs
//! Shared [`Channel`] trait + [`ChannelRegistry`] (B4-BE2).
//!
//! The two existing adapters ([`crate::service::run_qq_channel`] and
//! [`crate::telegram::run_telegram_channel`]) were wired ad-hoc from
//! `corlinman-gateway::main::maybe_spawn_*_channel`. This module extracts a
//! uniform contract so:
//!
//! 1. New inbound transports follow a single trait.
//! 2. The gateway spawns every enabled channel via one iteration
//!    ([`spawn_all`]) instead of bespoke helpers.
//! 3. Per-channel behaviour is unchanged — [`QqChannel`] / [`TelegramChannel`]
//!    are thin wrappers that forward [`Channel::run`] to the existing
//!    `run_*_channel` function bodies. No regression in the hot path.
//!
//! ## Why the trait surface is minimal
//!
//! The spec sketched `send` / `edit` / `typing` / `send_media` methods on the
//! trait, but today the reply path lives *inside* each adapter (it owns the
//! WS action channel / the Telegram reply mpsc). Exposing those as trait
//! methods now would require tearing out both adapters' internals — a change
//! the parent task explicitly forbids ("DO NOT touch qq/onebot.rs or
//! qq/service.rs internals — wrap them"). Instead the trait exposes only the
//! stable surface the gateway actually consumes (`id`, `enabled`, `run`);
//! outbound helpers can be added later with default `Unsupported` impls once
//! the adapters are refactored to thread sends through a shared channel.
//!
//! ## `ChannelError::Unsupported`
//!
//! Reserved for future outbound helpers on read-only channels. Kept here so
//! downstream crates can rely on the error variant name without churn when
//! those helpers land.

use std::sync::Arc;

use corlinman_core::config::Config;
use corlinman_gateway_api::ChatService;
use tokio::task::JoinHandle;
use tokio_util::sync::CancellationToken;

use crate::router::RateLimitHook;

/// Per-channel runtime handle shared by the gateway at spawn time.
///
/// `Clone` so `spawn_all` can hand each enabled adapter its own copy without
/// forcing callers to re-build it. All fields are `Arc`-wrapped or cheap to
/// clone.
#[derive(Clone)]
pub struct ChannelContext {
    /// Full config snapshot; each adapter pulls its own `[channels.*]`
    /// sub-section inside [`Channel::enabled`] + [`Channel::run`].
    pub config: Arc<Config>,
    /// Shared chat pipeline the gateway built on top of its `ChatBackend`.
    pub chat_service: Arc<dyn ChatService>,
    /// Default model id (`cfg.models.default`) for channels whose inbound
    /// events carry no model hint.
    pub model: String,
    /// Optional observation hook fired by the router each time a message is
    /// dropped by a rate-limit check. `None` in tests; `Some(..)` in prod
    /// where the gateway wires it to a Prometheus CounterVec.
    pub rate_limit_hook: Option<RateLimitHook>,
    /// Optional shared hook bus (B4-BE6). Threaded through to the router
    /// so rate-limit rejections surface on
    /// [`corlinman_hooks::HookEvent::RateLimitTriggered`] in addition to
    /// the legacy `rate_limit_hook` callback.
    pub hook_bus: Option<Arc<corlinman_hooks::HookBus>>,
}

/// Inbound channel adapter contract.
///
/// Implementations are constructed once at gateway boot and stored as
/// `Arc<dyn Channel>` in a [`ChannelRegistry`]. For each enabled channel the
/// gateway calls [`Channel::run`] on a dedicated task; the returned future
/// must honour the [`CancellationToken`] so shutdown drains in bounded time.
#[async_trait::async_trait]
pub trait Channel: Send + Sync + 'static {
    /// Short stable id (`"qq"`, `"telegram"`). Used for logging, metrics
    /// labels, and registry lookup.
    fn id(&self) -> &str;

    /// Human-readable name for admin UI / logs. Defaults to [`Self::id`].
    fn display_name(&self) -> &str {
        self.id()
    }

    /// Whether this channel is enabled for the given config snapshot.
    /// Called once per boot by [`spawn_all`].
    fn enabled(&self, cfg: &Config) -> bool;

    /// Run the adapter to completion or cancellation. `Ok(())` = graceful
    /// exit; `Err` = fatal configuration / transport error surfaced to the
    /// caller.
    async fn run(&self, ctx: ChannelContext, cancel: CancellationToken) -> anyhow::Result<()>;
}

/// Error surface reserved for future outbound helpers. Today only
/// [`ChannelError::Unsupported`] is used by mock impls in tests; the
/// adapters' internal outbound paths still live inside their `run_*_channel`
/// bodies (see module docs for rationale).
#[derive(Debug, thiserror::Error)]
pub enum ChannelError {
    /// This channel does not implement the requested operation.
    #[error("operation {0} not supported by this channel")]
    Unsupported(&'static str),
}

// ---------------------------------------------------------------------------
// Registry
// ---------------------------------------------------------------------------

/// Ordered set of [`Channel`] impls the gateway will try to spawn at boot.
///
/// Order is insertion order; [`ChannelRegistry::builtin`] preserves
/// `qq` → `telegram` (matches the pre-refactor `main.rs` call order so log
/// output stays identical).
#[derive(Default, Clone)]
pub struct ChannelRegistry {
    channels: Vec<Arc<dyn Channel>>,
}

impl ChannelRegistry {
    /// Empty registry. Mostly useful for tests.
    pub fn new() -> Self {
        Self::default()
    }

    /// Registry pre-populated with the built-in adapters: `qq`, `telegram`.
    pub fn builtin() -> Self {
        let mut r = Self::new();
        r.push(Arc::new(QqChannel));
        r.push(Arc::new(TelegramChannel));
        r
    }

    /// Append an adapter. External crates (future: Discord, Slack) can push
    /// their own impls into a registry before [`spawn_all`].
    pub fn push(&mut self, ch: Arc<dyn Channel>) {
        self.channels.push(ch);
    }

    /// Iterate registered adapters in insertion order.
    pub fn iter(&self) -> impl Iterator<Item = &Arc<dyn Channel>> {
        self.channels.iter()
    }

    /// Count of registered adapters (regardless of enabled state).
    pub fn len(&self) -> usize {
        self.channels.len()
    }

    /// True when no adapters are registered.
    pub fn is_empty(&self) -> bool {
        self.channels.is_empty()
    }
}

/// Spawn one task per enabled channel and return the join handles.
///
/// Disabled channels ([`Channel::enabled`] returns `false`) are skipped
/// without spawning; the returned `Vec` length matches the enabled count.
/// Each task's `JoinHandle` yields the channel's `run` result so the caller
/// can log per-channel failures on shutdown.
pub fn spawn_all(
    registry: &ChannelRegistry,
    ctx: ChannelContext,
    cancel: CancellationToken,
) -> Vec<JoinHandle<anyhow::Result<()>>> {
    registry
        .iter()
        .filter(|ch| ch.enabled(&ctx.config))
        .map(|ch| {
            let ch = Arc::clone(ch);
            let ctx = ctx.clone();
            let cancel = cancel.child_token();
            tokio::spawn(async move { ch.run(ctx, cancel).await })
        })
        .collect()
}

// ---------------------------------------------------------------------------
// Built-in adapters — thin wrappers around existing `run_*_channel` bodies.
// ---------------------------------------------------------------------------

/// QQ / OneBot v11 adapter. Forwards [`Channel::run`] to
/// [`crate::service::run_qq_channel`] so the runtime behaviour is bit-for-bit
/// identical to the pre-refactor `maybe_spawn_qq_channel` path.
pub struct QqChannel;

#[async_trait::async_trait]
impl Channel for QqChannel {
    fn id(&self) -> &str {
        "qq"
    }

    fn display_name(&self) -> &str {
        "QQ (OneBot v11)"
    }

    fn enabled(&self, cfg: &Config) -> bool {
        cfg.channels.qq.as_ref().map(|q| q.enabled).unwrap_or(false)
    }

    async fn run(&self, ctx: ChannelContext, cancel: CancellationToken) -> anyhow::Result<()> {
        // `enabled` is true so this `as_ref` always produces Some — but defend
        // against misuse by erroring instead of panicking.
        let qq_cfg = ctx
            .config
            .channels
            .qq
            .as_ref()
            .ok_or_else(|| anyhow::anyhow!("qq channel run() called but channels.qq is None"))?
            .clone();
        let params = crate::service::QqChannelParams {
            config: qq_cfg,
            model: ctx.model.clone(),
            chat_service: ctx.chat_service.clone(),
            rate_limit_hook: ctx.rate_limit_hook.clone(),
            hook_bus: ctx.hook_bus.clone(),
        };
        crate::service::run_qq_channel(params, cancel).await
    }
}

/// APNs (Apple Push Notification service) adapter — Phase 4 W3 C4
/// iter 8 stub.
///
/// The Swift macOS reference client (`apps/swift-mac/`) registers for
/// remote notifications and expects the gateway to push
/// `PushNotification` payloads via APNs HTTP/2. This stub reserves
/// the adapter slot in the [`ChannelRegistry`] without yet shipping
/// the HTTP/2 + JWT pipeline:
///
/// - [`Channel::enabled`] returns `false` until a future iter adds
///   `cfg.channels.apns` (config + provider URL + auth-key path).
///   That keeps `cargo build` green without forcing a config schema
///   change inside the C4 worktree.
/// - [`Channel::run`] is therefore never invoked at the moment; it
///   returns `Ok(())` defensively for the day someone wires the
///   config struct and the registration path lights up.
///
/// The dev-iteration loop uses a Unix-domain socket fallback driven
/// from `[channels.dev_push]` (also TBD); the Swift client reads
/// JSON lines off that socket, see
/// `apps/swift-mac/Sources/CorlinmanCore/PushReceiver.swift:200-220`.
pub struct ApnsChannel;

#[async_trait::async_trait]
impl Channel for ApnsChannel {
    fn id(&self) -> &str {
        "apns"
    }

    fn display_name(&self) -> &str {
        "APNs (stub)"
    }

    /// Disabled until the gateway grows `cfg.channels.apns`. Returning
    /// `false` here means [`spawn_all`] never invokes [`Channel::run`]
    /// on the stub, so the adapter is invisible at runtime.
    fn enabled(&self, _cfg: &Config) -> bool {
        false
    }

    /// Defensive no-op — `enabled` is `false` so this isn't reachable
    /// today. Kept here so the day someone flips the config flag the
    /// adapter doesn't panic; the actual HTTP/2 push path lands in a
    /// follow-on Wave 4 iter.
    async fn run(&self, _ctx: ChannelContext, _cancel: CancellationToken) -> anyhow::Result<()> {
        Ok(())
    }
}

/// Telegram adapter. Forwards to [`crate::telegram::run_telegram_channel`].
///
/// B4-BE1 is adding webhook handling inside the `telegram/` module in
/// parallel; once their public API stabilises this wrapper can switch
/// between webhook / long-poll based on `cfg.channels.telegram.mode`
/// without touching the trait surface.
pub struct TelegramChannel;

#[async_trait::async_trait]
impl Channel for TelegramChannel {
    fn id(&self) -> &str {
        "telegram"
    }

    fn display_name(&self) -> &str {
        "Telegram"
    }

    fn enabled(&self, cfg: &Config) -> bool {
        cfg.channels
            .telegram
            .as_ref()
            .map(|t| t.enabled)
            .unwrap_or(false)
    }

    async fn run(&self, ctx: ChannelContext, cancel: CancellationToken) -> anyhow::Result<()> {
        let tg_cfg = ctx
            .config
            .channels
            .telegram
            .as_ref()
            .ok_or_else(|| {
                anyhow::anyhow!("telegram channel run() called but channels.telegram is None")
            })?
            .clone();
        let params = crate::telegram::TelegramParams {
            config: tg_cfg,
            chat_service: ctx.chat_service.clone(),
            model: ctx.model.clone(),
        };
        crate::telegram::run_telegram_channel(params, cancel).await
    }
}

// ---------------------------------------------------------------------------
// Unit tests — integration-style tests that exercise enable/skip logic live
// in `tests/trait_impl.rs`.
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn registry_push_and_iter_preserves_order() {
        // Two no-op channels with distinct ids; iter should yield them in
        // insertion order so `ChannelRegistry::builtin` can rely on it.
        struct StubA;
        struct StubB;
        #[async_trait::async_trait]
        impl Channel for StubA {
            fn id(&self) -> &str {
                "a"
            }
            fn enabled(&self, _: &Config) -> bool {
                false
            }
            async fn run(&self, _: ChannelContext, _: CancellationToken) -> anyhow::Result<()> {
                Ok(())
            }
        }
        #[async_trait::async_trait]
        impl Channel for StubB {
            fn id(&self) -> &str {
                "b"
            }
            fn enabled(&self, _: &Config) -> bool {
                false
            }
            async fn run(&self, _: ChannelContext, _: CancellationToken) -> anyhow::Result<()> {
                Ok(())
            }
        }

        let mut r = ChannelRegistry::new();
        r.push(Arc::new(StubA));
        r.push(Arc::new(StubB));
        let ids: Vec<&str> = r.iter().map(|c| c.id()).collect();
        assert_eq!(ids, vec!["a", "b"]);
    }

    #[test]
    fn builtin_ordering_matches_pre_refactor_call_order() {
        let r = ChannelRegistry::builtin();
        let ids: Vec<&str> = r.iter().map(|c| c.id()).collect();
        assert_eq!(ids, vec!["qq", "telegram"]);
    }

    #[test]
    fn display_name_defaults_to_id() {
        struct S;
        #[async_trait::async_trait]
        impl Channel for S {
            fn id(&self) -> &str {
                "x"
            }
            fn enabled(&self, _: &Config) -> bool {
                false
            }
            async fn run(&self, _: ChannelContext, _: CancellationToken) -> anyhow::Result<()> {
                Ok(())
            }
        }
        assert_eq!(S.display_name(), "x");
    }
}

File: python/packages/corlinman-channels/src/corlinman_channels/channel.py
"""Shared :class:`Channel` Protocol + :class:`ChannelRegistry` (Python port).

Python port of ``rust/.../channel.rs``. The two existing adapters
(:func:`corlinman_channels.onebot.OneBotAdapter`,
:func:`corlinman_channels.telegram.TelegramAdapter`) were wired ad-hoc
from the gateway. This module extracts a uniform contract so:

1. New inbound transports follow a single Protocol.
2. The gateway spawns every enabled channel via one iteration
   (:func:`spawn_all`) instead of bespoke helpers.
3. Per-channel behaviour is unchanged — :class:`QqChannel` /
   :class:`TelegramChannel` are thin wrappers around the existing
   adapters.

## Why the contract is minimal

The Rust spec sketched ``send`` / ``edit`` / ``typing`` / ``send_media``
methods on the trait, but today the reply path lives *inside* each
adapter (OneBot owns the action channel; Telegram owns the reply mpsc).
Exposing those as Protocol methods now would require tearing out both
adapters' internals — a change the parent task explicitly forbids
("DO NOT touch onebot.py or telegram.py internals — wrap them"). So
the Protocol exposes only the stable surface the gateway actually
consumes (``id``, ``enabled``, ``run``); outbound helpers can be added
later with default :class:`ChannelError.Unsupported` impls.

## Deliberate deviations from Rust

- Rust uses ``Arc<Config>`` everywhere; Python uses a structural type
  (any object with the right attributes; tests pass plain
  ``SimpleNamespace``). Avoids a hard dependency on a corlinman-core
  Python port that doesn't exist yet.
- Rust returns ``JoinHandle<Result<()>>``; Python returns
  ``asyncio.Task[None]`` and the task itself surfaces ``Exception`` on
  ``await`` (matches asyncio idiom).
- Rust's ``CancellationToken`` becomes :class:`asyncio.Event` — call
  ``cancel.set()`` to request shutdown; the adapter's ``run``
  coroutine should observe and exit promptly.
- Rust's ``async_trait`` becomes ``typing.Protocol`` with an
  ``async def run``. Python's ABCs would also work; we pick Protocol
  for structural-typing parity with the rest of the package.
"""

from __future__ import annotations

import asyncio
from collections.abc import Iterator
from dataclasses import dataclass, field
from typing import Any, Protocol, runtime_checkable

from corlinman_channels.common import UnsupportedError

__all__ = [
    "ApnsChannel",
    "Channel",
    "ChannelContext",
    "ChannelError",
    "ChannelRegistry",
    "QqChannel",
    "TelegramChannel",
    "spawn_all",
]


# ---------------------------------------------------------------------------
# Error surface
# ---------------------------------------------------------------------------


class ChannelError(Exception):
    """Base error for channel-Protocol operations.

    Mirrors Rust's ``ChannelError`` enum (only the ``Unsupported``
    variant is in use today; see module docs for the rationale). Kept
    as a subclass-friendly base so future variants (``Transport``,
    ``Auth``) can land without changing the import surface.
    """

    @staticmethod
    def unsupported(operation: str) -> ChannelError:
        """Factory matching Rust ``ChannelError::Unsupported(op)``.

        Returns a :class:`UnsupportedError` (the existing
        :mod:`corlinman_channels.common` variant) with a stable
        ``"operation <op> not supported by this channel"`` message —
        downstream code can pattern-match on ``isinstance(..,
        UnsupportedError)`` or string-contains.
        """
        return UnsupportedError(f"operation {operation} not supported by this channel")


# ---------------------------------------------------------------------------
# Runtime context
# ---------------------------------------------------------------------------


@dataclass(slots=True)
class ChannelContext:
    """Per-channel runtime handle shared by the gateway at spawn time.

    Mirrors the Rust ``ChannelContext``. ``config`` is intentionally
    typed as ``Any`` because the corlinman-core Python port doesn't
    exist yet — callers pass a ``SimpleNamespace`` (or real config
    object) whose ``.channels.{qq,telegram,apns}`` attribute the
    built-in channels read in :meth:`Channel.enabled`.

    All fields are cheap to copy by reference; the dataclass is
    re-usable across every spawned task without an explicit ``clone()``.
    """

    config: Any
    """Full config snapshot; each adapter pulls its own ``channels.*``
    sub-section inside :meth:`Channel.enabled` + :meth:`Channel.run`."""

    chat_service: Any = None
    """Shared chat pipeline the gateway built on top of its
    ``ChatBackend``. Typed ``Any`` for the same reason as ``config``.
    May be ``None`` in trait-impl tests that never dispatch."""

    model: str = ""
    """Default model id for channels whose inbound events carry no
    model hint."""

    rate_limit_hook: Any = None
    """Optional observation hook fired by the router each time a
    message is dropped by a rate-limit check. ``None`` in tests;
    populated in prod where the gateway wires it to Prometheus."""

    hook_bus: Any = None
    """Optional shared :class:`corlinman_hooks.HookBus`. Threaded
    through to the router so rate-limit rejections surface on
    ``HookEvent.RateLimitTriggered`` in addition to the legacy
    callback."""


# ---------------------------------------------------------------------------
# Protocol
# ---------------------------------------------------------------------------


@runtime_checkable
class Channel(Protocol):
    """Inbound channel adapter contract.

    Implementations are constructed once at gateway boot and stored in
    a :class:`ChannelRegistry`. For each enabled channel, the gateway
    calls :meth:`run` on a dedicated task; the coroutine must honour
    the ``cancel`` :class:`asyncio.Event` so shutdown drains in bounded
    time.

    The Protocol is intentionally minimal — see the module docs for
    why outbound helpers aren't exposed here.
    """

    def id(self) -> str:
        """Short stable id (``"qq"``, ``"telegram"``). Used for
        logging, metric labels, and registry lookup."""
        ...

    def display_name(self) -> str:
        """Human-readable name for admin UI / logs. Defaults to
        :meth:`id`."""
        ...

    def enabled(self, cfg: Any) -> bool:
        """Whether this channel is enabled for ``cfg``. Called once
        per boot by :func:`spawn_all`."""
        ...

    async def run(self, ctx: ChannelContext, cancel: asyncio.Event) -> None:
        """Run the adapter to completion or cancellation. ``return``
        cleanly on graceful exit; raise for fatal configuration or
        transport errors so the caller can surface them."""
        ...


# ---------------------------------------------------------------------------
# Registry
# ---------------------------------------------------------------------------


@dataclass(slots=True)
class ChannelRegistry:
    """Ordered set of :class:`Channel` impls the gateway will try to
    spawn at boot.

    Order is insertion order; :meth:`builtin` preserves ``qq →
    telegram`` (matches the pre-refactor call order so log output
    stays identical to Rust).
    """

    _channels: list[Channel] = field(default_factory=list)

    @classmethod
    def builtin(cls) -> ChannelRegistry:
        """Registry pre-populated with the built-in adapters:
        ``qq`` → ``telegram``."""
        r = cls()
        r.push(QqChannel())
        r.push(TelegramChannel())
        return r

    def push(self, ch: Channel) -> None:
        """Append an adapter. External callers (future Discord / Slack
        adapters) can push their own impls before :func:`spawn_all`."""
        self._channels.append(ch)

    def iter(self) -> Iterator[Channel]:
        """Iterate registered adapters in insertion order."""
        return iter(self._channels)

    def __iter__(self) -> Iterator[Channel]:
        # Make the registry directly iterable too — feels natural in Python.
        return iter(self._channels)

    def __len__(self) -> int:
        return len(self._channels)

    def is_empty(self) -> bool:
        """True when no adapters are registered. Mirrors Rust
        ``ChannelRegistry::is_empty``."""
        return not self._channels


# ---------------------------------------------------------------------------
# spawn_all
# ---------------------------------------------------------------------------


def spawn_all(
    registry: ChannelRegistry,
    ctx: ChannelContext,
    cancel: asyncio.Event,
) -> list[asyncio.Task[None]]:
    """Spawn one task per enabled channel and return the task handles.

    Disabled channels (:meth:`Channel.enabled` returns ``False``) are
    skipped without spawning; the returned list's length matches the
    enabled count. Each task awaits the channel's ``run`` coroutine
    so the caller can ``await`` to surface per-channel failures on
    shutdown.

    Unlike Rust where each spawn gets a child cancel token, here we
    share the same :class:`asyncio.Event` because Python's asyncio
    doesn't have a built-in parent/child cancellation hierarchy — the
    common practice is to let the same event be the shutdown flag for
    every spawned task. Callers that need per-task cancellation can
    pass distinct events themselves.
    """
    tasks: list[asyncio.Task[None]] = []
    for ch in registry.iter():
        if not ch.enabled(ctx.config):
            continue
        # Snapshot the dataclass so each adapter has its own logical
        # copy (matches Rust ``ctx.clone()`` per spawn). The fields are
        # shared by reference; only the wrapper is new.
        local_ctx = ChannelContext(
            config=ctx.config,
            chat_service=ctx.chat_service,
            model=ctx.model,
            rate_limit_hook=ctx.rate_limit_hook,
            hook_bus=ctx.hook_bus,
        )
        task = asyncio.create_task(
            ch.run(local_ctx, cancel),
            name=f"channel-{ch.id()}",
        )
        tasks.append(task)
    return tasks


# ---------------------------------------------------------------------------
# Built-in adapters — thin wrappers
# ---------------------------------------------------------------------------


def _cfg_get(cfg: Any, *path: str) -> Any:
    """Walk ``cfg.path[0].path[1]...`` with ``None`` short-circuiting.

    The Rust crate reaches into a typed ``Config`` struct; the Python
    plane uses a duck-typed ``cfg`` so we walk it defensively here.
    Returns ``None`` on any missing/None step so the ``enabled``
    helpers can express their checks as
    ``_cfg_get(cfg, "channels", "qq", "enabled") is True``.
    """
    cur = cfg
    for step in path:
        if cur is None:
            return None
        # Support both attribute access (SimpleNamespace) and mapping
        # access (TOML dicts).
        cur = cur.get(step) if isinstance(cur, dict) else getattr(cur, step, None)
    return cur


class QqChannel:
    """QQ / OneBot v11 adapter wrapper.

    Forwards :meth:`run` to the QQ orchestration helper in
    :mod:`corlinman_channels.service`. Behaviour stays bit-for-bit
    compatible with the pre-refactor ``run_qq_channel`` call path.
    """

    def id(self) -> str:
        return "qq"

    def display_name(self) -> str:
        return "QQ (OneBot v11)"

    def enabled(self, cfg: Any) -> bool:
        return bool(_cfg_get(cfg, "channels", "qq", "enabled"))

    async def run(self, ctx: ChannelContext, cancel: asyncio.Event) -> None:
        # Lazy-import to break the import cycle:
        # ``service`` imports ``router`` imports nothing here, but the
        # built-in registry construction can run before ``service`` is
        # imported by the caller. Importing lazily keeps the channel
        # module standalone-importable.
        from corlinman_channels.service import QqChannelParams, run_qq_channel

        qq_cfg = _cfg_get(ctx.config, "channels", "qq")
        if qq_cfg is None:
            raise RuntimeError(
                "qq channel run() called but channels.qq is None"
            )
        params = QqChannelParams(
            config=qq_cfg,
            model=ctx.model,
            chat_service=ctx.chat_service,
            rate_limit_hook=ctx.rate_limit_hook,
            hook_bus=ctx.hook_bus,
        )
        await run_qq_channel(params, cancel)


class TelegramChannel:
    """Telegram long-poll adapter wrapper.

    Forwards :meth:`run` to the Telegram orchestration helper in
    :mod:`corlinman_channels.service`.
    """

    def id(self) -> str:
        return "telegram"

    def display_name(self) -> str:
        return "Telegram"

    def enabled(self, cfg: Any) -> bool:
        return bool(_cfg_get(cfg, "channels", "telegram", "enabled"))

    async def run(self, ctx: ChannelContext, cancel: asyncio.Event) -> None:
        from corlinman_channels.service import TelegramChannelParams, run_telegram_channel

        tg_cfg = _cfg_get(ctx.config, "channels", "telegram")
        if tg_cfg is None:
            raise RuntimeError(
                "telegram channel run() called but channels.telegram is None"
            )
        params = TelegramChannelParams(
            config=tg_cfg,
            model=ctx.model,
            chat_service=ctx.chat_service,
        )
        await run_telegram_channel(params, cancel)


class ApnsChannel:
    """APNs (Apple Push Notification service) stub.

    Mirrors the Rust ``ApnsChannel`` stub: reserves the registry slot
    without yet shipping the HTTP/2 + JWT pipeline. :meth:`enabled`
    returns ``False`` until the config grows ``channels.apns``, so
    :func:`spawn_all` never invokes :meth:`run` today. The defensive
    no-op body is kept for the day someone wires the config flag.
    """

    def id(self) -> str:
        return "apns"

    def display_name(self) -> str:
        return "APNs (stub)"

    def enabled(self, cfg: Any) -> bool:
        return False

    async def run(
        self,
        ctx: ChannelContext,
        cancel: asyncio.Event,
    ) -> None:
        """No-op until the APNs adapter actually lands. ``enabled``
        returns ``False`` so this is unreachable in practice."""
        return None
